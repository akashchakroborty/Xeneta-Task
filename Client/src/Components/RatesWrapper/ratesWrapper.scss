@import '../../App/app.variables.scss';

.ratesWrapper {
  margin: 0 4em;
  .ant-spin-spinning {
    display: flex;
    justify-content: center;
  }
  .ratesContainer > div {
    margin: 2.5em 0;
  }
  .ant-alert-message {
    color: $primary-color;
  }
}
