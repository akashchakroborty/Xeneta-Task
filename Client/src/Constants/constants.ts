export const SELECT_ORIGIN = 'Select Origin';
export const SELECT_DESTINATION = 'Select Destination';
export const EMPTY_STRING = '';
export const DEFAULT_PORTS = [];
export const DEFAULT_RATES = [];
export const DEFAULT_LOADING = false;
