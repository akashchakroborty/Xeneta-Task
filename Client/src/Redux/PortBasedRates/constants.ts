export const SET_PORTS_SUCCESS = 'SET_PORTS_SUCCESS';
export const SET_PORTS_FAILURE = 'SET_PORTS_FAILURE';
export const GET_PORTS_LOADING = 'GET_PORTS_LOADING';
export const UPDATE_ORIGIN = 'UPDATE_ORIGIN';
export const UPDATE_DESTINATION = 'UPDATE_DESTINATION';
