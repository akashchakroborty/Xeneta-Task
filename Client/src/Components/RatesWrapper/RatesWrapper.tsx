import React from 'react';
import { Alert, RadioChangeEvent, Spin } from 'antd';
import Chart from '../../CommonComponents/Chart/Chart';
import { Rates } from '../../Redux/Types/portBasedRatesTypes';
import RadioGroup from '../../CommonComponents/RadioGroup/RadioGroup';
import { NO_DATA_MESSAGE, RATE_TYPE_MAPPER } from '../../Constants/constants';

import './ratesWrapper.scss';

export interface RatesWrapperProps {
  rates: Rates[];
  isRatesLoading: boolean;
  isGetRateClicked: boolean;
  selectedRateType: string;
  updateRateType: (selectedRateType: string) => void;
}
const RatesWrapper: React.FC<RatesWrapperProps> = ({
  rates,
  isRatesLoading,
  isGetRateClicked,
  selectedRateType,
  updateRateType,
}) => {
  const handleRadioChange = (e: RadioChangeEvent) => {
    updateRateType(e.target.value);
  };

  return (
    <div className="ratesWrapper">
      {isRatesLoading ? (
        <Spin className="spinner" size="large" />
      ) : (
        <>
          {rates.length > 0 ? (
            <div className="ratesContainer">
              <RadioGroup
                selectedValue={selectedRateType}
                data={RATE_TYPE_MAPPER}
                onChange={handleRadioChange}
              />
              <Chart
                data={rates}
                dataKeyX="day"
                dataKeyY={selectedRateType}
                width={1300}
                height={400}
              />
            </div>
          ) : (
            <>
              {isGetRateClicked && (
                <Alert className="alert" message={NO_DATA_MESSAGE} type="warning" showIcon />
              )}
            </>
          )}
        </>
      )}
    </div>
  );
};

export default RatesWrapper;
