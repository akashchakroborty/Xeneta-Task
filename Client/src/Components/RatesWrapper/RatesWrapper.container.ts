import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { updateRateType } from '../../Redux/PortBasedRates/actionCreators';
import {
  selectIsGetRateClicked,
  selectIsRatesLoading,
  selectRates,
  selectRateType,
} from '../../Redux/PortBasedRates/selector';
import { AppState } from '../../Redux/store';
import { PortBasedRatesActionTypes } from '../../Redux/Types/portBasedRatesActionTypes';
import RatesWrapper from './RatesWrapper';

const mapStateToProps = (state: AppState) => ({
  rates: selectRates(state),
  isRatesLoading: selectIsRatesLoading(state),
  selectedRateType: selectRateType(state),
  isGetRateClicked: selectIsGetRateClicked(state),
});

const mapDispatchToProps = (dispatch: ThunkDispatch<AppState, void, PortBasedRatesActionTypes>) => {
  return {
    updateRateType: (selectedRateType: string) => {
      dispatch(updateRateType(selectedRateType));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(RatesWrapper);
