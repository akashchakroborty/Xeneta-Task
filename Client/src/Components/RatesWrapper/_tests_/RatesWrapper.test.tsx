import React from 'react';
import { shallow } from 'enzyme';
import RatesWrapper, { RatesWrapperProps } from '../RatesWrapper';
import { DEFAULT_RATES, DEFAULT_RATE_TYPE } from '../../../Constants/constants';
import { MOCK_RATES_DATA } from '../../../Constants/testData';

describe('should test the OriginDestination Component', () => {
  const mockProps: RatesWrapperProps = {
    rates: MOCK_RATES_DATA,
    isRatesLoading: false,
    isGetRateClicked: false,
    selectedRateType: DEFAULT_RATE_TYPE,
    updateRateType: jest.fn(),
  };
  test('should render the Component', () => {
    const component = shallow(<RatesWrapper {...mockProps} />);
    expect(component).toMatchSnapshot();
  });
  test('should test handleClick function', () => {
    const wrapper = shallow(<RatesWrapper {...mockProps} />);
    const event = { target: { value: 'Mock Value' } };
    wrapper.find('RadioGroupComponent').simulate('change', event);
    expect(mockProps.updateRateType).toHaveBeenCalled();
  });
  test('spinner should exist when isRatesLoading is true', () => {
    const newMockProps = { ...mockProps, isRatesLoading: true };
    const wrapper = shallow(<RatesWrapper {...newMockProps} />);
    expect(wrapper.find('.spinner').exists()).toBe(true);
  });
  test('Should show warning notification when isGetRateClicked is true and rates is Empty.', () => {
    const newMockProps = { ...mockProps, isGetRateClicked: true, rates: DEFAULT_RATES };
    const wrapper = shallow(<RatesWrapper {...newMockProps} />);
    expect(wrapper.find('.alert').exists()).toBe(true);
  });
});
