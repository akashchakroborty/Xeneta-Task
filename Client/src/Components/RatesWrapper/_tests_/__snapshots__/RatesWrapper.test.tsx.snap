// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the OriginDestination Component should render the Component 1`] = `
<div
  className="ratesWrapper"
>
  <div
    className="ratesContainer"
  >
    <RadioGroupComponent
      data={
        Map {
          "Market Average" => "mean",
          "Market Low" => "low",
          "Market High" => "high",
        }
      }
      onChange={[Function]}
      selectedValue="mean"
    />
    <ChartComponent
      data={
        Array [
          Object {
            "day": "2021-01-01",
            "high": 2436,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-02",
            "high": 2436,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-03",
            "high": 2436,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-04",
            "high": 2435,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-05",
            "high": 2427,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-06",
            "high": 2431,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-07",
            "high": 2437,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-08",
            "high": 2438,
            "low": 1036,
            "mean": 1619,
          },
          Object {
            "day": "2021-01-09",
            "high": 2438,
            "low": 1036,
            "mean": 1619,
          },
          Object {
            "day": "2021-01-10",
            "high": 2438,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-11",
            "high": 2436,
            "low": 1037,
            "mean": 1615,
          },
          Object {
            "day": "2021-01-12",
            "high": 2436,
            "low": 1036,
            "mean": 1614,
          },
          Object {
            "day": "2021-01-13",
            "high": 2436,
            "low": 1036,
            "mean": 1614,
          },
          Object {
            "day": "2021-01-14",
            "high": 2436,
            "low": 1036,
            "mean": 1614,
          },
          Object {
            "day": "2021-01-15",
            "high": 2438,
            "low": 1036,
            "mean": 1614,
          },
          Object {
            "day": "2021-01-16",
            "high": 2437,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-17",
            "high": 2437,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-18",
            "high": 2437,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-19",
            "high": 2437,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-20",
            "high": 2437,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-21",
            "high": 2436,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-22",
            "high": 2434,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-23",
            "high": 2434,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-24",
            "high": 2434,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-25",
            "high": 2435,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-26",
            "high": 2436,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-27",
            "high": 2437,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-28",
            "high": 2438,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-29",
            "high": 2435,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-30",
            "high": 2435,
            "low": 1037,
            "mean": 1611,
          },
          Object {
            "day": "2021-01-31",
            "high": 2435,
            "low": 1037,
            "mean": 1611,
          },
        ]
      }
      dataKeyX="day"
      dataKeyY="mean"
      height={400}
      width={1300}
    />
  </div>
</div>
`;
